### ISSUE TITLE
[Typescript] declaration file missing

### ISSUE THREAD

**Issue by romainbellande** [joepuzzo/informed#37]
I don't find informed definition file in the provided package and @types/informed doesn't exist yet. Could you migrate @types/react-form with the changes ?

**Comment by joepuzzo** [comment]
This one is always fun. I dont ever use typescript. Someone created a pull request for react form last time and thats how it got types. I would be more than happy to approve a PR for this

**Comment by pure180** [comment]
Hey Guys,

finished a rough TypeScript definition for this package `informed@0.1.4` right know. Mostly everything is working.
Will Keep it on a regular git account and commit it to the definitely typed library when everything has been tested successfully. 

Feel free to test it and leave any comments or enhancements. Thank you!

Install it by using yarn:
`yarn add --dev https://ventzke-media@bitbucket.org/ventzkemediaberlin/types-informed.git#0.1.40`

**Comment by joepuzzo** [comment]
@pure180 when you get that to a stable point put out a PR!

**Comment by romainbellande** [comment]
@pure180 I tried it and it works well for now ! I will leave a comment if anything goes wrong :) thanks a lot you rock !

**Comment by pure180** [comment]
@joepuzzo Depends on how many people help to improve it. Bur for now everything, except a few things, are working fine. One more thing saw the `Field` class which is getting exported. but cant find any explanation for it, is there any?

@romainbellande You are welcome! Need it anyways too. :)

**Comment by joepuzzo** [comment]
@pure180 yeah so the `Field` class was used in `react-form` and i replaced it with `asField` HOC in this lib.. i kept exporting it for backwards compatibility but am unsure if im going to keep it or not.

**Comment by pure180** [comment]
@joepuzzo ok cool, it is typed but will remove it when you do so. Already saw that you added the async stuff in `1.5.0` will implement it too.

**Comment by pure180** [comment]
@joepuzzo Implemented it in a pretty complex, dynamic and fully typed registration form, including almost every component that informed offers (except custom inputs) and it is working like a charm, even with `informed@1.5.1`. Would like to have some people looking at it, but for me it could get merged.

**Comment by pure180** [comment]
@joepuzzo Do you want to include the type definitions to your library, or shall i publish it at definitely typed?

**Comment by joepuzzo** [comment]
You can create a pr for this and i will merge.. might be cleaner to keep it with the repo

**Comment by joepuzzo** [comment]
After doing some reading i actually think it would be better to do the latter and publish the types under the @types scope.

**Comment by FredyC** [comment]
@pure180 Hey, thanks for the typings. Would be lovely to get it more generic though. For example, it could look like this which gives much stronger types in overall. It's quite useless to be getting `any` everywhere. I have a data types coming from fully typed GraphQL interface, so I would like to use that. I made myself a similar modification for react-form and it worked very nicely.

```ts
  export interface FormProps<T = any> {
    onChange?: (formState: FormState<T>) => void;
    onSubmit?: (values: FormValues<T>) => void;
    onValueChange?: (value: FormValues<T>) => void;
    ... // update other types accordingly
  }
```

**Comment by FredyC** [comment]
@joepuzzo Personally I like better when typings and included in the package. It's definitely easier to use when I don't need to look for other extra packages to install and use it right away. On the other hand, there are many smart people on DefinitelyTyped and they can help to make types much better in overall.

**Comment by pure180** [comment]
Hello @FredyC ... already did exactly what you are suggesting, even fixed some wrong types too. So I'm completely with you, using strong typed properties. Pushing it right now. Will also add the async stuff, that @joepuzzo implemented recently, in the next days.

**Comment by FredyC** [comment]
@pure180 Well, if current version in the repo is the latest, then it's not exactly what I had in mind. The `FormProps` are still without generics. The example I posted above is relying on inference. If I pass in `initialValues` to the component, remaining props will be strongly typed as well because T will be inferred from it. Generic FormValues/FormState is the start, but it can be better :)

**Comment by pure180** [comment]
Sure, you could create a PR, would do it either. But I'm on business vacations right now, so could take some time till I'm able to do it.

**Comment by neezer** [comment]
Any progress on this? Just discovered `informed`—it looks great!—but types are a must for me. 

Gonna give @pure180's repo a spin for now and report any issues I find. 😄

**Comment by rscotten** [comment]
@pure180 So far I noticed that the <Form> element does not accept the "id" prop.

```
[ts] Property 'id' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Form> & Readonly<{ children?: ReactNode; }> & Readonly<FormProps<{}> & { children?: string | number | boolean | {} | ReactElement<any> | ReactNodeArray | ReactPortal | Element | ((props: FormClassContext) => RenderReturn) | never[] | null | undefined; }>'.
```

**Comment by yozi-developer** [comment]
@rscotten, @pure180  Could it be related to https://bitbucket.org/ventzkemediaberlin/types-informed/commits/22063ed7d80041c5a87785513e200d5954a4073d ? 

Also, there was a mistake before (HTMLFontElement instead of HTMLFormElement)

**Comment by yozi-developer** [comment]
Btw, I think definitions of the `asField` is not correct. 

`function asField<T, P>(Component: React.RefForwardingComponent<T, P>): AsFieldReturn<P>;`
means the wrapped component must be a function that accepts 2 arguments: props and ref.  But in fact `Component` is not a `RefForwardingComponent`, it is a `React.ComponentType`  that accepts `forwardedRef`

**Comment by pure180** [comment]
Hey @all

Thanks for your feedback, I'll implement it this weekend, when I find some spare time. Anyway you could also create a pull request if you have time to develop this, would be nice. Thank you guys. :)

**Comment by robbyemmert** [comment]
Has there been any progress on this? Types are a must for me too. I'm super excited, this is the react form package I've been waiting for.

Here's my 2 cents about the typedefs:
- Keeping the definitions with the package is definitely better. It keeps the definitions in sync. Also, if you ever wanted to convert the project to Typescript, it could generate type definitions automatically.
- Be sure you include a reference to the type definitions in a 'types' property on your package.json. That way, TS will pick up the defs automatically on import.

Thanks!

**Comment by dzwiedziu-nkg** [comment]
1. When you publish it as NPM package?
2. Please add `BasicRadioGroup`

**Comment by JaffParker** [comment]
I'm new to Typescript and unfortunately I'm having issues with your declaration package...

Here's what I'm trying to do:

```tsx
import React from 'react'
import {withFormState, FormState} from 'informed'
import {Button, ButtonProps} from 'reactstrap'

interface SubmitButtonProps extends ButtonProps {
  formState: FormState
  submitting: boolean
}

export const SubmitButton = withFormState<Button, SubmitButtonProps>(
  ({
    formState: {pristine, invalid},
    disabled = false, submitting = false, children, ref,
    ...props
  }) =>
    <Button type="submit" color="success" disabled={pristine || invalid || submitting || disabled}
            className={submitting ? 'progress-bar-striped progress-bar-animated' : ''}
            {...props}>
      {children}
    </Button>
)
```

Here's what happens when I try to use it:

![screenshot from 2018-10-09 17-00-37](https://user-images.githubusercontent.com/10477757/46698384-ef259600-cbe4-11e8-8bd2-f39bf80d2082.png)

It is as though it simply ignores the prop type that I defined. I cannot find what exactly can be causing it from the declaration files.

Another problem I've faced is that `Field` component does not accept `validateOn*` props, even though that is where they're supposed to go. `Basic*` components accept those props, but they simply pass them through to HTML elements which makes React throw a warning and obviously doesn't achieve the necessary functionality. It's something that should be corrected.

**Comment by JaffParker** [comment]
@pure180 I think I figured what the problem might be:

```ts
export function withFormState<T, P = {}>(Component: FormHocComponent<T, P>): React.ComponentType<React.ClassAttributes<P>>;
```

`React.ClassAttributes` does not accept a prop types definition, but an element to pass to Ref. Thus you should not be using `P` as it's generic type argument, but `T` instead. So here's what it probably should look like:

```ts
export function withFormState<T, P = {}>(Component: FormHocComponent<T, P>): React.ComponentType<React.ClassAttributes<T>&P>;
```

I haven't tried it yet, I'm not in front of my computer, I'll test it later. Ideally also, it should be typing `formState` and `formApi`, but I had to do it manually...

**Comment by JaffParker** [comment]
@pure180 I tried my fixes and they worked. I forked your repo, made the changes, but unfortunately, you blocked pull requests. Please, allow them so you can merge my changes.

**Comment by pure180** [comment]
@JaffParker, Sorry I didn't found the settings to unblock pull requests from forks, bitbucket sucks, lol. 

I'll create a new repo here on GitHub, so that you can either post you code here, create a gist or fork the new repo. Thanks and sorry for your efforts. 

@all others sorry I am still quite busy lately. Will continue on the types as soon as I got spare time.

Thanks for you feedback!

**Comment by FrimJo** [comment]
Thank you for an awesome library! I see that a new version containing typescript definition files has been released, but when running `yarn add informed` I do not get any definition files with my 1.10.8 version.
![image](https://user-images.githubusercontent.com/4533443/48265325-87a17700-e42c-11e8-89a3-afda43306f7b.png)

**Comment by FredyC** [comment]
@FrimJo I don't think there is an official release just yet. https://bitbucket.org/ventzkemediaberlin/types-informed/ is all you got right now.

**Comment by FrimJo** [comment]
@FredyC Thnx for the url, but I believe the `index.d.ts` file added to this project works better, but not fully. https://github.com/joepuzzo/informed/blob/master/index.d.ts

**Comment by Verbon** [comment]
Hi guys,

It seems like @FrimJo is correct: typescript definitions were committed and merged into master.

But it seems that they just don't reach the 'dist' folder (which is actually a published package) after the build process.
![image](https://user-images.githubusercontent.com/9406367/48269869-ac581780-e449-11e8-8e1d-48b8dfba37e2.png)

`"files"` attribute determines what we get when installing a package via yarn/npm (https://docs.npmjs.com/files/package.json#files).

So, I see 2 ways to resolve this issue:
1. Modify `"files"` attribute of `package.json` file to include `index.d.ts` file.
2. Modify the build process, so `index.d.ts` file is copied to the `dist` folder as well.

@joepuzzo Please, take a look at this and choose the option which suits you best :)

**Comment by JaffParker** [comment]
@Verbon that's my bad, I created another PR to fix that. @joepuzzo pls merge https://github.com/joepuzzo/informed/pull/107

**Comment by gsinovsky** [comment]
Hi guys, I'm trying to pass `forwadedRef` to a `Text` component, but the property doesn't exist on the current TypeScript definition.

**Comment by JaffParker** [comment]
@gsinovsky that's my bad, I made a minimal declaration, followed the docs and I missed `forwardedRef` unfortunately. Do you think you can add it yourself?

**Comment by gsinovsky** [comment]
@JaffParker sure

**Comment by gsinovsky** [comment]
@JaffParker [here it is](https://github.com/joepuzzo/informed/pull/110), check it out if you can plese.

**Comment by Verbon** [comment]
@JaffParker @gsinovsky @FrimJo
Guys, I have a great news for all of you - there's a very cool library out there called [Formik](https://jaredpalmer.com/formik/docs/overview). It's dead simple, but it does just a lot of annoying stuff including handling nested objects, arrays, validation (integrated with [Yup](https://github.com/jquense/yup)) etc.

**But the greatest thing here - it's already fully written in TypeScript**. So forget about the types headache. The types will be always correct and updated automatically with every new library release.

I've already tried it - and it's the greatest form library for React applications I've ever seen so far.
So, I would suggest to forget about informed, and switch to Formik instead.

"Build React forms, without tears" (c) Jared Palmer at Formik :)

**Comment by FredyC** [comment]
@Verbon Are you being serious? It's missing even basic things like `onChange` on the `Form` or even giving access to imperative formApi. Any async validation? Documentation is really sparse considering how old the project is. Nah, thanks...

**Comment by Verbon** [comment]
@FredyC Before making any conclusions, please carefully read the docs first. It has everything you need, you just didn't even try to read. It has a different API, so you won't find any `onChange` and `formApi` props. You would use `values` prop of the FormikProp instead, and any of the `set%something%` imperative methods accordingly.

Please, use informed if you like it so much, but it will be an endless headache for you if you're a TypeScript user.

**Comment by FrimJo** [comment]
I love this project and I like the structure of it, and I believe that fixing the above mentioned issues will satesify my TS needs.

**Comment by gsinovsky** [comment]
> I wrote this lib and even I just wish the react guys would make a form lib so we could all stop using different ones 🤣

https://xkcd.com/927/

@joepuzzo

**Comment by joepuzzo** [comment]
Hahaha. I have a backround in ipv6 testing and trust me when i say standards are a beautiful thing! React and forms will get there one day... maybe

**Comment by Ricki-BumbleDev** [comment]
Looks like types are included in the package now, right? https://github.com/joepuzzo/informed/blob/master/index.d.ts
This could be closed now, I guess...

**Comment by joepuzzo** [comment]
Needs to be updated for V2.. i dont have time right now but is somone wants to update and make a PR then they are more than welcome

**Comment by JaffParker** [comment]
I'm working on it slowly, but surely. Should be done over the weekend, hopefully
